// cot-lib/dist/clients/layer-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Observable_1 = require("rxjs/Observable");
var axios_1 = require("axios");
var util_1 = require("../util");
var internal_utils_1 = require("../internal-utils");
var LayerClient = /** @class */ (function () {
    function LayerClient(host, axiosInstance, cityId) {
        this.host = host;
        this.ax = axiosInstance || axios_1.default.create();
        this.cityId = cityId;
    }
    /**
     * Get layer data of the requested metric.
     * @param metric Metric of the data for the layer
     */
    LayerClient.prototype.getLayer = function (metric) {
        var that = this;
        return {
            getHour: function (hourTS) {
                var hourStr = util_1.Util.Time.getHourStr(hourTS);
                var url = internal_utils_1.url(that.host, 'layers', that.cityId, metric, hourStr);
                return internal_utils_1.InternalUtils.getAsTemporalPage(url, that.ax);
            },
            getHourRange: function (fromTS, toTS) {
                return that.getLayer(metric).getHour(fromTS).expand(function (tpage) {
                    var url = tpage.link.next;
                    if (tpage.link.self.endsWith(util_1.Util.Time.getHourStr(toTS))) {
                        return Observable_1.Observable.empty();
                    }
                    else {
                        return internal_utils_1.InternalUtils.getAsTemporalPage(that.host + url, that.ax);
                    }
                });
            },
            getHourRecap: function (fromTS, toTS) {
                return that.getLayer(metric).getHourRange(fromTS, toTS)
                    .map(function (o) { return o.data; })
                    .reduce(function (acc, curr) { return internal_utils_1.InternalUtils.mergeDataLayerData(acc, curr); });
            },
            getHourRecapOf: function () {
                var hourTS = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    hourTS[_i] = arguments[_i];
                }
                return Observable_1.Observable.from(hourTS)
                    .flatMap(function (ts) { return that.getLayer(metric).getHour(ts); })
                    .filter(function (o) { return !o.data.nodata; })
                    .map(function (o) { return o.data; })
                    .reduce(function (acc, curr) { return internal_utils_1.InternalUtils.mergeDataLayerData(acc, curr); });
            },
            getDay: function (dayTS) {
                var day = util_1.Util.Time.getDayStr(dayTS);
                var url = internal_utils_1.url(that.host, 'layers', that.cityId, metric, day);
                return internal_utils_1.InternalUtils.getAsTemporalPage(url, that.ax);
            },
            getDayRange: function (fromTS, toTS) {
                return that.getLayer(metric).getDay(fromTS).expand(function (tpage) {
                    var url = tpage.link.next;
                    if (tpage.link.self.endsWith(util_1.Util.Time.getDayStr(toTS))) {
                        return Observable_1.Observable.empty();
                    }
                    else {
                        return internal_utils_1.InternalUtils.getAsTemporalPage(that.host + url, that.ax);
                    }
                });
            },
            getDayRecap: function (fromTS, toTS) {
                return that.getLayer(metric).getDayRange(fromTS, toTS)
                    .map(function (o) { return o.data; })
                    .reduce(function (acc, curr) { return internal_utils_1.InternalUtils.mergeDataLayerData(acc, curr); });
            },
            getDayRecapOf: function () {
                var dayTS = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    dayTS[_i] = arguments[_i];
                }
                return Observable_1.Observable.from(dayTS)
                    .flatMap(function (ts) { return that.getLayer(metric).getDay(ts); })
                    .map(function (o) { return o.data; })
                    .reduce(function (acc, curr) { return internal_utils_1.InternalUtils.mergeDataLayerData(acc, curr); });
            },
        };
    };
    return LayerClient;
}());
exports.LayerClient = LayerClient;
